luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -18,52 +18,67 @@ local fallDamageMagnificationFactor = 14 --the multiplier by which default engin
-local maxImpulseThreshold = 0.08 --incease this to make units move less from impulse. This defines the max impulse damage a unit can take expressed in a percentage multiplier
-local cooldownsFrameThreshold = Game.gameSpeed --this is how long in frames before a unit's maximum impulse limit resets.
+local maxImpulseProportion = 0.04 --incease this to make units move less from impulse. This defines the max impulse damage allowed a unit can take relative to its mass.
 
-local spGetUnitVelocity = Spring.GetUnitVelocity
 local spGetUnitHealth = Spring.GetUnitHealth
-local mathmin = math.min
+local spGetUnitVelocity = Spring.GetUnitVelocity
+local spSetUnitVelocity = Spring.SetUnitVelocity
+local mathMin = math.min
+local mathMax = math.max
+local mathCeil = math.ceil
 
 local fallDamageMultipliers = {}
 local unitsMaxImpulse = {}
 local weaponDefIDImpulses = {}
-local overImpulseCooldowns = {}
-local transportingUnits = {}
+local transportedUnits = {}
 local unitMasses = {}
-
-local currentModulusFrame = 0
+local weaponDefIgnored = {}
+local unitInertiaCheckFlags = {} -- track inertia via velocity (units have constant mass)
+local gameFrame = 0
+local velocityWatchDuration = 3
 
 for unitDefID, unitDef in ipairs(UnitDefs) do
 	local fallDamageMultiplier = unitDef.customParams.fall_damage_multiplier or 1.0
-	fallDamageMultipliers[unitDefID] = fallDamageMultiplier*fallDamageMagnificationFactor
-	unitsMaxImpulse[unitDefID] = unitDef.mass/maxImpulseThreshold
+	fallDamageMultipliers[unitDefID] = fallDamageMultiplier * fallDamageMagnificationFactor
+	unitsMaxImpulse[unitDefID] = unitDef.mass / maxImpulseProportion
 	unitMasses[unitDefID] = unitDef.mass
 end
 
+for name, weaponDefID in pairs(Game.envDamageTypes or {}) do
+	weaponDefIgnored[weaponDefID] = true
+end
+
 for weaponDefID, wDef in ipairs(WeaponDefs) do
 	if wDef.damages and wDef.damages.impulseBoost and wDef.damages.impulseFactor then
 		weaponDefIDImpulses[weaponDefID] = {impulseboost = wDef.damages.impulseBoost, impulsefactor = wDef.damages.impulseFactor}
 	end
-end
 
-local function calculateImpulseData(unitDefID, damage, weaponDefID)
-	local impulse = (damage + weaponDefIDImpulses[weaponDefID].impulseboost) * weaponDefIDImpulses[weaponDefID].impulsefactor
-	local maxImpulse = unitsMaxImpulse[unitDefID]
-	local impulseMultiplier = mathmin(maxImpulse/impulse, 1)
-	return impulse, impulseMultiplier
-end
-
-local function isUnitBeingTransported(unitID)
-	for transportID, transportedID in pairs(transportingUnits) do
-		if transportingUnits[transportID][unitID] then
-			return true
+	
+	local function maxDamage(damages)
+		local damage = damages[0]
+		for i = 1, #damages do
+			damage = mathMax(damages[i], damage)
 		end
+		return damage
+	end
+	
+	--generate list of exempted weapons to improve performance
+	if wDef.damages and wDef.damages.impulseFactor == 0 or
+		(wDef.damages.impulseFactor < 0.15 and wDef.damages.impulseBoost < maxDamage(wDef.damages) * 0.2) then
```

Comment:
Where do these magic numbers come from? What do they represent?

Reply:
ignore weapons that don't have significant impulse impact relative to damage. This boosts performance. 
Making these magic numbers variables doesn't make intent any more clear IMO.
Impulsefactor and impulseboost are usually set to .123 each when impulse isn't desired in weapondefs

Reply:
Extract to two config variables, `minImpulseFactor` and `minBoostToDamageRatio`. But doesn't this make any `impulseFactor` and `impulseBoost` entries in weapon defs no-op if they don't meet these thresholds? If that is the case, a warning should be added to the log when a weapon is detected with values that don't meet the thresholds.

Reply:
Dunno what no-op is. It's the normal state of things for some weapons to have such pitiful impulse that they aren't worth even performing an impulsecap calculation. This saves performance. you'd be warning "hey dude, LLT's too too small impulse damage!"... which is normal and good.

no logging required.

I don't agree with your code style choice here, but I'll do it anyways.

Reply:
no-op = no-operation, code that runs but produces no effect

I'm saying that the values in the unit defs should be set to 0, so that they reflect reality. Unit def files should be WYSIWYG, we have real problems with the value in the file not being the value reflected in game.